File: JanperCab/5-JanperCab/ClientApp/src/app/_models/delivery-sheet/DeliverySheetDto.ts
import { EnquiryForSheetDto } from './../enquiry/EnquiryForSheetDto';
import { DeliveryMethodEnum } from './../../_enums/DeliveryMethodEnum';
import { Type } from 'class-transformer';

export abstract class DeliverySheetDto {
  $type: string;
  id: number;
  deliveryMethod: DeliveryMethodEnum;
  createdDate: Date;
  lockedDate: Date;
  completedDate: Date;

  @Type(() => EnquiryForSheetDto)
  enquiriesForSheet: EnquiryForSheetDto[];

  constructor() {
    this.$type = '';
  }

  get isEditable(): boolean {
    return !this.lockedDate && !this.completedDate;
  }

  abstract getBarcodePrefix(): string;
  abstract getDescription(): string;
}

File: JanperCab/3-Application/Dtos/DeliverySheet/DeliverySheetDto.cs
using _1_Domain.Enum;
using _3_Application.Dtos.Enquiry;
using System;
using System.Collections.Generic;

namespace _3_Application.Dtos.DeliverySheet
{
    public abstract class DeliverySheetDto
    {
        public int Id { get; set; }

        public DeliveryMethodEnum DeliveryMethod { get; set; }

        public DateTime CreatedDate { get; set; }

        public DateTime? LockedDate { get; set; }

        public DateTime? CompletedDate { get; set; }

        public List<EnquiryForSheetDto> EnquiriesForSheet { get; set; }


        protected DeliverySheetDto()
        {
            EnquiriesForSheet = new List<EnquiryForSheetDto>();
        }
    }
}